### PR TITLE
Display time remaining until video goes live

### LINKED ISSUE #2391: [Feature Request]: Show time left until livestream/premiere starts
### Guidelines

- [X] I have searched the [issue tracker](https://github.com/FreeTubeApp/FreeTube/issues) for a feature request that matches the one I want to file, without success.
- [X] I have searched the [documentation](https://docs.freetubeapp.io/) for information that matches the description of the feature request I want to file, without success.

### Problem Description

Youtube displays how much time is left until a live stream / premiere goes live along with the exact date and time.
This makes it easier to tell when the video starts. This is very useful when it's scheduled a couple of days in the future.

Freetube currently does not display this.

### Proposed Solution

Display how much time is left until the live stream or video premiere goes live above the exact premiere datetime. 
This would make it easier to tell at a glance when a live stream/premiere will start and would be similar to how it looks on YouTube.

AFAIK Youtube only updates this text when the page is refreshed so the Freetube implementation doesn't have to be any more complicated than that.

### Alternatives Considered

I've not considered any other alternatives.

### Issue Labels

display more information to user, improvement to existing feature

### Additional Information

How it looks on youtube:
2022-07-09_T21:51:46

### PR DESCRIPTION
**Pull request by Aiz0**
---
Display time remaining until video goes live
---

**Important note**
We may remove your pull request if you do not use this provided PR template correctly.

**Pull Request Type**
Please select what type of pull request this is:
- [ ] Bugfix
- [x] Feature Implementation
- [ ] Documentation
- [ ] Other

**Related issue**
closes #2391
alleviates the timezone issue on flatpak #1429

**Description**
Display how much time is left until the live stream or video premiere goes live above the exact premiere datetime.
This is static and will not update automatically as the premiere time gets closer.

The time left will be displayed in the format "Premieres in X U" where X is a number and U is a time unit such as hours, minutes, or days. I use `Math.floor()` on the number so it's always a whole number. From what I have seen this is what YouTube seems to do.

When to switch from hours to minutes to seconds remaining is based on how I understand YouTube does it.

**Screenshots (if appropriate)**

Before:
2022-08-20_T19:20:23

After:
2022-08-20_T19:48:09

with custom date
2022-08-20_T18:27:46
2022-08-20_T18:28:13
2022-08-20_T18:28:57
2022-08-20_T18:29:22
2022-08-20_T18:30:41

**Testing (for code that is not small enough to be easily understandable)**
Has this pull request been tested?
Please describe shortly how you tested it and whether there are any ramifications remaining. 

I used upcoming livestreams from YouTube [live channel](https://www.youtube.com/channel/UC4R8DWoMoI7CAwX8_LjQHig) and compared the times.
I also tested with a custom date instead of `upcomingTimestamp` to see if all types of time remaining where displayed correctly.

**Desktop (please complete the following information):**
 - OS Version: [e.g. 22]
 - FreeTube version: [e.g. 0.8]

**Additional context**
Is it possible to differentiate between live streams and premieres?
If so, then it might be good to have separate strings for them. Such as "Live in" and "Premieres In"

Sometimes YouTube seems to display time remaining like this.
2022-08-20_T16:35:10
This seems to happen when there is less than 60 minutes left.
I don't know if they do this on all videos now, but that format should only be considered if we update the time automatically

Since I am new to FreeTube development I would appreciate help and feedback.

### CODE REVIEW

**Comment by ChunkyProgrammer** [comment]
Thanks for opening a pr! Would you be able to lint it using `npm run lint` and applying the fixes?

File: src/renderer/views/Watch/Watch.js [hunk 1]
```diff
@@ -395,8 +396,38 @@ export default Vue.extend({
...
+              let timeUnit
+              if (upcomingTimeLeft <= 0) {
+                this.upcomingTimeLeft = this.$t('Video.Starting Soon')
+              } else {
+                upcomingTimeLeft = upcomingTimeLeft / 1000
+                timeUnit = 'Second'
+                if (upcomingTimeLeft > 60) {
+                  upcomingTimeLeft = upcomingTimeLeft / 60
+                  timeUnit = 'Minute'
+                  if (upcomingTimeLeft > 120) {
```
Comment (ChunkyProgrammer) [inline src/renderer/views/Watch/Watch.js:410]: Haven't tested yey, but shouldn't this be 60 instead of 120? (Correct me if I'm wrong)

Comment (Aiz0) [reply src/renderer/views/Watch/Watch.js:410]: AFAIK youtube switches to showing time left in minutes at 120 minutes remaining. I tried to replicate how they choose to display it.

File: src/renderer/views/Watch/Watch.js [hunk 2]
```diff
@@ -395,8 +396,38 @@ export default Vue.extend({
...
+                  }
+                }
+                upcomingTimeLeft = Math.floor(upcomingTimeLeft)
+                if (upcomingTimeLeft !== 1) {
+                  timeUnit = timeUnit + 's'
+                }
+                timeUnit = this.$t('Video.Published.' + timeUnit)
+
+                // TODO a template might be needed here but idk how they work
+                this.upcomingTimeLeft = upcomingTimeLeft + ' ' + timeUnit.toLowerCase()
```
Comment (PikachuEXE) [inline src/renderer/views/Watch/Watch.js:426]: Better temp variable scoping, flatten nested code, rename temp variables, use string intepolation
```suggestion
              if (upcomingTimeLeft <= 0) {
                this.upcomingTimeLeft = this.$t('Video.Starting Soon')
              } else {
                // Convert from ms to second
                upcomingTimeLeft = upcomingTimeLeft / 1000
                let timeUnitI18nPartialKey = 'Second'

                if (upcomingTimeLeft > 60) {
                  upcomingTimeLeft = upcomingTimeLeft / 60
                  timeUnitI18nPartialKey = 'Minute'
                }

	            // Youtube switches to showing time left in minutes at 120 minutes remaining
	            if (timeUnitI18nPartialKey === 'Minute' && upcomingTimeLeft > 120) {
	              upcomingTimeLeft = upcomingTimeLeft / 60
	              timeUnitI18nPartialKey = 'Hour'
	            }
	              
	            if (timeUnitI18nPartialKey === 'Hour' && upcomingTimeLeft > 24) {
	              upcomingTimeLeft = upcomingTimeLeft / 24
	              timeUnitI18nPartialKey = 'Day'
	            }
                
                // Value after decimal not to be displayed
                // e.g. > 2 days = display as `2 days`
                upcomingTimeLeft = Math.floor(upcomingTimeLeft)
                if (upcomingTimeLeft !== 1) {
                  timeUnitI18nPartialKey = timeUnitI18nPartialKey + 's'
                }
                const timeUnitTranslated = this.$t(`Video.Published.${timeUnitI18nPartialKey}`).toLowerCase()

                // TODO a I18n entry for time format might be needed here
                this.upcomingTimeLeft = `${upcomingTimeLeft} ${timeUnitTranslated}`
```

File: src/renderer/views/Watch/Watch.vue [hunk 3]
```diff
@@ -54,7 +54,7 @@
               v-if="upcomingTimestamp !== null"
               class="premiereText"
             >
-              {{ $t("Video.Premieres on") }}:
+              {{ $t("Video.Premieres in") }} {{ upcomingTimeLeft }}
```
Comment (PikachuEXE) [inline src/renderer/views/Watch/Watch.vue:57]: So it's possible to have `Premieres in Starting soon`?

**Review by PikachuEXE — CHANGES_REQUESTED** [review]

File: src/renderer/views/Watch/Watch.js (hunk 2 above)
Comment (Aiz0) [reply src/renderer/views/Watch/Watch.js]: Thanks,
I had trouble with naming the time unit variable.

Should string interpolation be used when adding 's' to timeUnitI18nPartialKey, or is that not preferred there?

File: src/renderer/views/Watch/Watch.vue (hunk 3 above)
Comment (Aiz0) [reply src/renderer/views/Watch/Watch.vue:57]: I guess I forgot to test that, that would be possible and would look weird.

After testing it with a video that was about to go live I noticed that when time has passed the scheduled timestamp but the video is not yet live then `upcomingTimeStamp` will be null.
Now I have to wonder if it's possible to even see it in normal use cases.

**Comment by Aiz0** [comment]
Seeing as this is doesn't update automatically would this be preferred instead of showing "premiers in x seconds"? 
2022-08-23_T20:54:19

File: src/renderer/views/Watch/Watch.js (hunk 2 above)
Comment (PikachuEXE) [reply src/renderer/views/Watch/Watch.js]: Ya that can be done as well I am just being lazy :P

**Comment by PikachuEXE** [comment]
> Seeing as this is doesn't update automatically would this be preferred instead of showing "premiers in x seconds"?
Yes

**Comment by PikachuEXE** [comment]
Let me know when final changes made will put review comment (should pass is no new issue found)

**Comment by Aiz0** [comment]
I've made the final changes now.

**Review by PikachuEXE — COMMENTED** [review]
Code looks good, testing locally later

**Comment by PikachuEXE** [comment]
I can't find any video on https://www.youtube.com/channel/UC4R8DWoMoI7CAwX8_LjQHig
Any other channels can be used to test?

**Comment by absidue** [comment]
> I can't find any video on https://www.youtube.com/channel/UC4R8DWoMoI7CAwX8_LjQHig Any other channels can be used to test?

You have to open the link on YouTube, it's a special page (so it doesn't work in FreeTube) that shows videos that are currently live and ones that have recently been live.

**Review by PikachuEXE — APPROVED** [review]
Tested with a few about to live videos in https://www.youtube.com/channel/UC4R8DWoMoI7CAwX8_LjQHig

**Review by ChunkyProgrammer — APPROVED** [review]
LGTM

**Review by absidue — APPROVED** [review]

**Pull request merged**